Predict probabilities

nbokulich opened the pull request:
All classifiers have a new output: predicted class probabilities. In other words, the predicted likelihood that each test sample belongs to each sample class. Useful for all sorts of things.

Specify a new format: `ProbabilitiesFormat`

Added tests for the format, and for replication of expected probabilities.

thermokarst commented on q2_sample_classifier/_format.py:
```diff
@@ -163,3 +163,15 @@ def _validate_(self, level):
     ImportanceFormat)
+
+
+class ProbabilitiesFormat(ImportanceFormat):
```
How about instead we create a new base class and have both of these formats inherit from it?

nbokulich replied:
👍

thermokarst commented on q2_sample_classifier/_transformer.py:
```diff
@@ -65,13 +66,14 @@ def _4(data: pd.Series) -> (PredictionsFormat):
     with ff.open() as fh:
         df = _read_dataframe(fh)
-        return df.iloc[:, 0]
+        return pd.to_numeric(df.iloc[:, 0], errors='ignore')
```
What was going on that prompted this change? Do we need to add a test case?

thermokarst replied:
Oh is this the same changeset as https://github.com/qiime2/q2-sample-classifier/pull/158? If so, you can address my questions there.

nbokulich replied:
yes sorry need to rebase (forgot I had pulled in those changes here), will respond over there

thermokarst commented on q2_sample_classifier/classify.py:
```diff
@@ -262,20 +266,29 @@ def predict_base(table, sample_estimator, n_jobs):
 def predict_regression(table: biom.Table, sample_estimator: Pipeline,
                        n_jobs: int = defaults['n_jobs']) -> pd.Series:
-    return predict_base(table, sample_estimator, n_jobs)
+    # we only return the predictions, not the probabilities, which are empty
```
:cool:

nbokulich commented:
thanks @thermokarst ! I added the base class `MultiColumnNumericFormat` per your request.

I rebased in an attempt to clean up the history but looks like that failed...

thermokarst commented on q2_sample_classifier/_format.py:
```diff
@@ -120,7 +120,7 @@ def _validate_(self, level):
 
 
-class ImportanceFormat(model.TextFileFormat):
+class MultiColumnNumericFormat(model.TextFileFormat):
```
```suggestion
class _MultiColumnNumericFormat(model.TextFileFormat):
```

Let's mark this as "private", that way an endeavoring developer doesn't get too carried away with this format.

nbokulich replied:
👍

nbokulich replied:
thanks for breaking my tests, dude 😆
